cpp/src/arrow/compute/kernels/take-test.cc

```diff
@@ -0,0 +1,175 @@
+                  TakeOptions options, const std::string& expected) {
+    TestTakeKernel<ArrowType>::AssertTake(type_singleton(), values, indices, options,
+                                          expected);
+  }
+  std::shared_ptr<DataType> type_singleton() {
+    return TypeTraits<ArrowType>::type_singleton();
+  }
+};
+
+TYPED_TEST_CASE(TestTakeKernelWithNumeric, NumericArrowTypes);
+TYPED_TEST(TestTakeKernelWithNumeric, TakeNumeric) {
+  TakeOptions options;
+  this->AssertTake("[7, 8, 9]", "[0, 1, 0]", options, "[7, 8, 7]");
+  this->AssertTake("[null, 8, 9]", "[0, 1, 0]", options, "[null, 8, null]");
+  this->AssertTake("[7, 8, 9]", "[null, 1, 0]", options, "[null, 8, 7]");
+
+  std::shared_ptr<Array> arr;
+  ASSERT_RAISES(Invalid, this->Take(this->type_singleton(), "[7, 8, 9]", "[0, 9, 0]",
+                                    options, &arr));
+
+  options.out_of_bounds = TakeOptions::TO_NULL;
+  this->AssertTake("[7, 8, 9]", "[0, 9, 0]", options, "[7, null, 7]");
+}
+
+class TestTakeKernelWithString : public TestTakeKernel<StringType> {
+ protected:
+  void AssertTake(const std::string& values, const std::string& indices,
+                  TakeOptions options, const std::string& expected) {
+    TestTakeKernel<StringType>::AssertTake(utf8(), values, indices, options, expected);
+  }
+  void AssertTakeDictionary(const std::string& dictionary_values,
+                            const std::string& dictionary_indices,
+                            const std::string& indices, TakeOptions options,
+                            const std::string& expected_indices) {
+    auto type = dictionary(int8(), ArrayFromJSON(utf8(), dictionary_values));
+    std::shared_ptr<Array> values, actual, expected;
+    ASSERT_OK(DictionaryArray::FromArrays(type, ArrayFromJSON(int8(), dictionary_indices),
+                                          &values));
+    ASSERT_OK(DictionaryArray::FromArrays(type, ArrayFromJSON(int8(), expected_indices),
+                                          &expected));
+    auto take_indices = ArrayFromJSON(int8(), indices);
+    this->AssertTakeArrays(values, take_indices, options, expected);
+  }
+};
+
+TEST_F(TestTakeKernelWithString, TakeString) {
+  TakeOptions options;
+  this->AssertTake(R"(["a", "b", "c"])", "[0, 1, 0]", options, R"(["a", "b", "a"])");
+  this->AssertTake(R"([null, "b", "c"])", "[0, 1, 0]", options, "[null, \"b\", null]");
+  this->AssertTake(R"(["a", "b", "c"])", "[null, 1, 0]", options, R"([null, "b", "a"])");
+
+  std::shared_ptr<Array> arr;
+  ASSERT_RAISES(Invalid,
+                this->Take(utf8(), R"(["a", "b", "c"])", "[0, 9, 0]", options, &arr));
+
+  options.out_of_bounds = TakeOptions::TO_NULL;
+  this->AssertTake(R"(["a", "b", "c"])", "[0, 9, 0]", options, R"(["a", null, "a"])");
+}
+
+TEST_F(TestTakeKernelWithString, TakeDictionary) {
+  TakeOptions options;
+  auto dict = R"(["a", "b", "c", "d", "e"])";
+  this->AssertTakeDictionary(dict, "[0, 1, 4]", "[0, 1, 0]", options, "[0, 1, 0]");
```

Comment:
Please use a more interesting example: with `[0, 1, 4]`, `index == value` for the first two values and you only take those ones. Something like `[3, 4, 2]` would be fine.

Reply:
@pitrou how's it now?

Reply:
Better, thanks!